// effect/dist/cjs/ConfigSecret.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.value = exports.unsafeWipe = exports.make = exports.isConfigSecret = exports.fromString = exports.fromChunk = exports.ConfigSecretTypeId = void 0;
var internal = /*#__PURE__*/_interopRequireWildcard( /*#__PURE__*/require("./internal/configSecret.js"));
function _getRequireWildcardCache(e) {
  if ("function" != typeof WeakMap) return null;
  var r = new WeakMap(),
    t = new WeakMap();
  return (_getRequireWildcardCache = function (e) {
    return e ? t : r;
  })(e);
}
function _interopRequireWildcard(e, r) {
  if (!r && e && e.__esModule) return e;
  if (null === e || "object" != typeof e && "function" != typeof e) return {
    default: e
  };
  var t = _getRequireWildcardCache(r);
  if (t && t.has(e)) return t.get(e);
  var n = {
      __proto__: null
    },
    a = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var u in e) if ("default" !== u && Object.prototype.hasOwnProperty.call(e, u)) {
    var i = a ? Object.getOwnPropertyDescriptor(e, u) : null;
    i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u];
  }
  return n.default = e, t && t.set(e, n), n;
}
/**
 * @since 2.0.0
 * @category symbols
 */
const ConfigSecretTypeId = exports.ConfigSecretTypeId = internal.ConfigSecretTypeId;
/**
 * @since 2.0.0
 * @category refinements
 */
const isConfigSecret = exports.isConfigSecret = internal.isConfigSecret;
/**
 * @since 2.0.0
 * @category constructors
 */
const make = exports.make = internal.make;
/**
 * @since 2.0.0
 * @category constructors
 */
const fromChunk = exports.fromChunk = internal.fromChunk;
/**
 * @since 2.0.0
 * @category constructors
 */
const fromString = exports.fromString = internal.fromString;
/**
 * @since 2.0.0
 * @category getters
 */
const value = exports.value = internal.value;
/**
 * @since 2.0.0
 * @category unsafe
 */
const unsafeWipe = exports.unsafeWipe = internal.unsafeWipe;
